// @google-labs/breadboard/dist/src/traversal/result.js
/**
 * @license
 * Copyright 2023 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { MachineEdgeState } from "./state.js";
export class MachineResult {
    descriptor;
    inputs;
    missingInputs;
    opportunities;
    newOpportunities;
    state;
    outputsPromise;
    pendingOutputs;
    constructor(descriptor, inputs, missingInputs, opportunities, newOpportunities, state, pendingOutputs) {
        this.descriptor = descriptor;
        this.inputs = inputs;
        this.missingInputs = missingInputs;
        this.opportunities = opportunities;
        this.newOpportunities = newOpportunities;
        this.state = state;
        this.pendingOutputs = pendingOutputs;
    }
    /**
     * `true` if the machine decided that the node should be skipped, rather than
     * visited.
     */
    get skip() {
        return this.missingInputs.length > 0;
    }
    static fromObject(o) {
        const edgeState = new MachineEdgeState();
        edgeState.constants = o.state.constants;
        edgeState.state = o.state.state;
        return new MachineResult(o.descriptor, o.inputs, o.missingInputs, o.opportunities, o.newOpportunities, edgeState, o.pendingOutputs);
    }
}
